### PR TITLE
Validate: check conflicting ipfamily and pods / services subnet

### PR DESCRIPTION
**Pull request by fedepaol**
When choosing a single stack ip family, we must ensure that the pod subnet / service subnet belong to the same family.

Fixes https://github.com/kubernetes-sigs/kind/issues/2468

### CODE REVIEW

File: pkg/internal/apis/config/validate.go [hunk 1]
```diff
@@ -168,6 +168,10 @@ func validateSubnets(subnetStr string, dualstack bool) error {
 	// if not DualStack only one CIDR allowed
 	case !dualstack && len(subnets) > 1:
 		allErrs = append(allErrs, errors.New("only one CIDR allowed for single-stack networking"))
+	case ipFamily == IPv4Family && subnets[0].IP.To4() == nil:
+		allErrs = append(allErrs, errors.New("IPv4 family but IPv6 CIDR"))
+	case ipFamily == IPv6Family && subnets[0].IP.To4() != nil:
+		allErrs = append(allErrs, errors.New("IPv6 family but IPv4 CIDR"))
```
Comment (neolit123) [inline pkg/internal/apis/config/validate.go:174]: I would reword the errors to:
> expected IPvX CIDR for IPvX family

Comment (fedepaol) [reply pkg/internal/apis/config/validate.go]: done, thanks!

File: pkg/internal/apis/config/validate.go [hunk 2]
```diff
@@ -168,6 +168,10 @@ func validateSubnets(subnetStr string, dualstack bool) error {
 	// if not DualStack only one CIDR allowed
 	case !dualstack && len(subnets) > 1:
 		allErrs = append(allErrs, errors.New("only one CIDR allowed for single-stack networking"))
+	case ipFamily == IPv4Family && subnets[0].IP.To4() == nil:
```
Comment (aojea) [inline pkg/internal/apis/config/validate.go:171]: can subnets be 0?
maybe we should be paranoid and consider that to avoid a panic here

Comment (fedepaol) [reply pkg/internal/apis/config/validate.go:174]: Fair enough, I added a case len == 0 on top of the switch

File: pkg/internal/apis/config/validate.go [hunk 3]
```diff
@@ -153,7 +152,11 @@ func validateSubnets(subnetStr string, dualstack bool) error {
 		subnets = append(subnets, cidr)
 	}
 
+	dualstack := ipFamily == DualStackFamily
 	switch {
+	// if no subnets are defined
+	case len(subnets) == 0:
+		allErrs = append(allErrs, errors.New("no subnets defined"))
```
Comment (aojea) [inline pkg/internal/apis/config/validate.go:159]: IIRC defaulting is before validation, but in case of having zero here, is this an error or is a "no need validation"  @BenTheElder ?

**Comment by aojea** [comment]
/ok-to-test
just one question about the case with no subnets

File: pkg/internal/apis/config/validate.go (hunk 3 above)
Comment (BenTheElder) [reply pkg/internal/apis/config/validate.go:159]: we only added this case because https://github.com/kubernetes-sigs/kind/pull/2473/files#r714772513 :-)

IIRC defaulting does come first so maybe we should be throwing a "this can't happen" type error (maybe even panic?) /shrug

Comment (aojea) [reply pkg/internal/apis/config/validate.go:159]: We can have an error and add a test for protecting of future changes that may remove the defaulting, if we panic and we don't test it we never now until someone hits the panic

@fedepaol can you add a test that it fails without subnets with a comment explaining that this should never happen for future us?

Comment (fedepaol) [reply pkg/internal/apis/config/validate.go:159]: Will do

**Comment by aojea** [comment]
/lgtm
/approve
Thank you very much

**Comment by BenTheElder** [comment]
thanks!
